bip78/src/receiver/state.rs

```diff
@@ -0,0 +1,263 @@
+use bitcoin::util::psbt::PartiallySignedTransaction;
+use std::marker::PhantomData;
+use std::convert::{TryFrom, TryInto};
+use crate::bitcoin::{Transaction, Script};
+use crate::psbt::Psbt;
+use bitcoin::OutPoint;
+
+#[derive(Clone, Debug)]
+pub struct PsbtState<S> {
+    psbt: PartiallySignedTransaction,
+    state: S
+}
+
+#[derive(Debug)]
+pub enum PsbtError {
+    UnequalInputCounts { tx_ins: usize, psbt_ins: usize, },
+    UnequalOutputCounts { tx_outs: usize, psbt_outs: usize, },
+    Todo,
+}
+
+pub trait Next<S> {
+    fn next(self) -> PsbtState<S>;
+}
+
+pub trait TryNext<S> {
+    fn try_next(self) -> Result<PsbtState<S>, PsbtError>;
+}
+
+/// Validate a [`PartiallySignedTransaction`] by checking transaction inputs number are equal to the
+/// psbt inputs and the same for outputs. Note deserialization already have this guarantee (TODO verify)
+#[derive(Clone, Debug)]
+pub struct Validated;
+
+///
+#[derive(Clone, Debug, Default)]
+pub struct MaybeUnbroadcastable {
+    is_broadcastable: bool,
+}
+
+///
+#[derive(Clone, Debug, Default)]
+pub struct MaybeInputsOwned {
+    are_previous_script_pubkey_not_mine: bool,
+}
+
+///
+#[derive(Clone, Debug, Default)]
+pub struct MaybePrevoutsSeen {
+    are_prevouts_never_seen: bool,
+}
+
+///
+#[derive(Clone, Debug)]
+pub struct Proposal;
+
+impl TryFrom<PartiallySignedTransaction> for PsbtState<Validated> {
+    type Error = PsbtError;
+
+    fn try_from(psbt: PartiallySignedTransaction) -> Result<Self, Self::Error> {
+        let tx_ins = psbt.global.unsigned_tx.input.len();
+        let psbt_ins = psbt.inputs.len();
+        let tx_outs = psbt.global.unsigned_tx.output.len();
+        let psbt_outs = psbt.outputs.len();
+
+        if psbt_ins != tx_ins {
+            Err(PsbtError::UnequalInputCounts { tx_ins, psbt_ins, })
+        } else if psbt_outs != tx_outs {
+            Err(PsbtError::UnequalOutputCounts { tx_outs, psbt_outs, })
+        } else {
```

Comment:
We have a [newtype](https://github.com/Kixunil/payjoin/blob/master/bip78/src/psbt.rs#L27) for this